sycl/doc/extensions/ExtendedAtomics/SYCL_INTEL_extended_atomics.asciidoc

```diff
@@ -113,389 +115,399 @@ In SYCL 1.2.1, support for floating-point types is limited to the +load+, +store
+  Integral operator^=(Integral) const noexcept;
 
-|+template <typename pointerT> atomic(multi_ptr<pointerT, addressSpace> ptr)+
-|Permitted data types for +pointerT+ are any valid scalar data type which is the same size in bytes as +T+.  Constructs an instance of SYCL +atomic+ which is associated with the pointer +ptr+, converted to a pointer of data type +T+.
-|===
+};
 
-==== To:
+// Partial specialization for floating-point types
+template <memory_order DefaultOrder, access::address_space Space>
+class atomic_ref<Floating, DefaultOrder, Space> {
 
-|===
-|Constructor|Description
+  /* All other members from atomic_ref<T> are available */
 
-|+atomic(multi_ptr<T, addressSpace> ptr)+
-|Constructs an instance of SYCL +atomic+ which is associated with the pointer +ptr+.
-|===
+  using difference_type = value_type;
 
-=== Modify Table 4.101
+  Floating fetch_add(Floating operand,
+    memory_order order = default_read_modify_write_order);
 
-==== Add:
+  Floating fetch_sub(Floating operand,
+    memory_order order = default_read_modify_write_order);
 
-|===
-|Member function|Description
-|+bool is_lock_free() const+
-|Return +true+ if the atomic operations provided by this SYCL +atomic+ are lock-free.
+  Floating fetch_min(Floating operand,
+    memory_order order = default_read_modify_write_order);
 
-|+bool compare_exchange_weak(T &expected, T desired, memory_order order = memory_order::relaxed)+
-|Atomically compares the value at the address of the +multi_ptr+ associated with this SYCL +atomic+ against the value of +expected+.  If the values are equal attempts to replaces value at address of the +multi_ptr+ associated with this SYCL +atomic+ with the value of +desired+, otherwise assigns the original value at the address of the +multi_ptr+ associated with this SYCL +atomic+ to +expected+.  Returns +true+ if the comparison operation and replacement operation were successful.  The memory order of this atomic operation must be +memory_order::relaxed+ for both success and fail.
+  Floating fetch_max(Floating operand,
+    memory_order order = default_read_modify_write_order);
 
-|+bool compare_exchange_weak(T &expected, T desired, memory_order order = memory_order::relaxed)+
-|Equivalent to +compare_exchange_weak(expected, desired, order, order)+.
+  Floating operator++(int) const noexcept;
+  Floating operator--(int) const noexcept;
+  Floating operator++() const noexcept;
+  Floating operator--() const noexcept;
+  Floating operator+=(Floating) const noexcept;
+  Floating operator-=(Floating) const noexcept;
 
-|+bool compare_exchange_strong(T &expected, T desired, memory_order order = memory_order::relaxed)+
-|Equivalent to +compare_exchange_strong(expected, desired, order, order)+.
+};
 
-|+operator T() const+
-|Equivalent to +load()+.
+// Partial specialization for pointers
+template <template T, memory_order DefaultOrder, access::address_space Space>
+class atomic_ref<T*, DefaultOrder, Space> {
```

Comment:
Probably I missing something, but I don't fully understand how it will work:

```c++
template <memory_order DefaultOrder, access::address_space Space>
class atomic_ref<Integral, DefaultOrder, Space> { /* ... */ };

template <template T, memory_order DefaultOrder, access::address_space Space>
class atomic_ref<T*, DefaultOrder, Space> { /* ... */ };

atomic_ref<int/*, ... */> a; // which one will be created: for pointer or for integer?
```

Reply:
An `atomic_ref<int>` is a reference to an `int` declared in the corresponding address space.  An `atomic_ref<int*>` is a reference to an `int*`.

```c++
cgh.submit([&](handler& cgh) {
  auto A = accessor<int, 1, access::mode::read_write>(bufA, cgh);
  auto B = accessor<int*, 1, access::mode::read_write>(bufB, cgh);
  cgh.parallel_for(..., [=](...) {
    atomic_ref<int, ...>(A[i])++; // Integral specialization
    atomic_ref<int*, ...>(B[i])++; // Pointer specialization
  });
});
```

Reply:
Ok, I got it. Made the following code snippet for myself for education purposes: https://godbolt.org/z/txwfzF